Sources/SwiftCompilerPlugin/Macros.swift

```diff
@@ -0,0 +1,254 @@
+  func expandFreestandingMacro(
+    macro: PluginMessage.MacroReference,
+    discriminator: String,
+    expandingSyntax: PluginMessage.Syntax
+  ) throws {
+    let sourceManager = SourceManager()
+    let context = PluginMacroExpansionContext(
+      sourceManager: sourceManager,
+      discriminator: discriminator
+    )
+
+    let syntax = sourceManager.add(expandingSyntax)
+    guard let macroSyntax = syntax.asProtocol(FreestandingMacroExpansionSyntax.self) else {
+      let diag = PluginMessage.Diagnostic(
+        from: .init(node: syntax, message: MacroExpansionError.freestandingMacroSyntaxIsNotMacro),
+        in: sourceManager
+      )
+      try pluginHostConnection.sendMessage(
+        .expandFreestandingMacroResult(
+          expandedSource: "",
+          diagnostics: [diag]
+        )
+      )
+      return
+    }
+
+    let macroDef = self.resolveMacro(macro)
+    ///
+    if let exprMacroDef = macroDef as? ExpressionMacro.Type {
+      let rewritten = try exprMacroDef.expansion(of: macroSyntax, in: context)
+      let diagnostics = context.diagnostics.map {
+        PluginMessage.Diagnostic(from: $0, in: sourceManager)
+      }
+
+      let resultMessage = PluginToHostMessage.expandFreestandingMacroResult(
+        expandedSource: rewritten.description,
+        diagnostics: diagnostics
+      )
+
+      try pluginHostConnection.sendMessage(resultMessage)
+      return
+    }
+
+    if let declMacroDef = macroDef as? DeclarationMacro.Type {
+      let rewritten = try declMacroDef.expansion(of: macroSyntax, in: context)
+      let diagnostics = context.diagnostics.map {
+        PluginMessage.Diagnostic(from: $0, in: sourceManager)
+      }
+
+      let resultMessage = PluginToHostMessage.expandFreestandingMacroResult(
+        expandedSource: rewritten.description,
+        diagnostics: diagnostics
+      )
+
+      try pluginHostConnection.sendMessage(resultMessage)
+      return
+    }
+
+    let diag = PluginMessage.Diagnostic(
+      from: Diagnostic(node: syntax, message: MacroExpansionError.macroTypeNotFound),
+      in: sourceManager
+    )
+
+    try pluginHostConnection.sendMessage(
+      .expandFreestandingMacroResult(expandedSource: "", diagnostics: [diag])
+    )
+  }
+
+  /// Expand `@attached(XXX)` macros.
+  func expandAttachedMacro(
+    macro: PluginMessage.MacroReference,
+    macroRole: PluginMessage.MacroRole,
+    discriminator: String,
+    customAttributeSyntax: PluginMessage.Syntax,
+    declSyntax: PluginMessage.Syntax,
+    parentDeclSyntax: PluginMessage.Syntax?
+  ) throws {
+    let sourceManager = SourceManager()
+    let context = PluginMacroExpansionContext(
+      sourceManager: sourceManager,
+      discriminator: discriminator
+    )
+
+    guard let macroDefinition = resolveMacro(macro) else {
+      fatalError("macro type not found: \(macro.moduleName).\(macro.typeName)")
+    }
+
+    let customAttributeNode = sourceManager.add(customAttributeSyntax).cast(AttributeSyntax.self)
```

Comment:
Mostly out of interest: Why is this named `customAttributeSyntax`. The `CustomAttributeSyntax` type in SwiftSyntax has been renamed to `AttributeSyntax`.

Reply:
Just be cause ASTGen call it `customAttrNode` https://github.com/apple/swift/blob/a0e1810f627add51fbbe3ddf2c3ce9a9c6e3afdc/lib/ASTGen/Sources/ASTGen/Macros.swift#L438-L445

Reply:
I would assume that’s because ASTGen might pre-date the rename `CustomAttributeSyntax` -> `AttributeSyntax`. I think we should just call it `attributeNode` here (and ideally also rename `ASTGen` but that’s not as important IMO).

Reply:
Sounds good. Since this name is also used in `PluginMessage`, we need to change ASTGen too. I will open a PR to `swift` repo.